core/src/main/scala/zquery/DataSource.scala

```diff
@@ -0,0 +1,149 @@
+package zquery
+
+import zio.ZIO
+
+/**
+ * A `DataSource[R, E, A]` is capable of executing requests of type `A` that
+ * require an environment `R` and may fail with an `E`.
+ *
+ * Data sources must implement the method `run` which takes a collection of
+ * requests and returns an effect with a `CompletedRequestMap` containing a
+ * mapping from requests to results. Because `run` is parameterized on a
+ * collection of requests rather than a single request, data sources have the
+ * ability to introspect on all the requests being executed in parallel and
+ * optimize the query.
+ *
+ * Data sources will typically be parameterized on a subtype of `Request[A]`,
+ * though that is not strictly necessarily as long as the data source can map
+ * the request type to a `Request[A]`. Data sources can then pattern match on
+ * the collection of requests to determine the information requested, execute
+ * the query, and place the results into the `CompletedRequestsMap` using
+ * [[CompletedRequestMap.empty]] and [[CompletedRequestMap.insert]]. Data
+ * sources must provide requests for all results received or fail with an `E`.
+ * Failure to do so will result in a query dieing with a `QueryFailure`.
+ *
+ * Data sources are guaranteed that all requests will be unique when they are
+ * called by `ZQuery`.
+ */
+trait DataSource[-R, +E, -A] {
+  def dataSource: DataSource.Service[R, E, A]
+}
+
+object DataSource {
+
+  trait Service[-R, +E, -A] { self =>
+
+    val identifier: String
+
+    def run(requests: Iterable[A]): ZIO[R, E, CompletedRequestMap]
```

Comment:
Might be nice to add a comment mentioning this will NOT be called with an empty list. So you don't need to check if the list is empty in your implementation of this function.

Reply:
Done.